File: customer_m/modules/workbench_board.py
"""Project board aggregation for meeting and daily follow-up views."""

from __future__ import annotations

import sqlite3

from ..database import row_to_dict
from .workbench_common import (
    DONE_STATUSES,
    _done_sql,
    _enrich_project_summaries,
    _placeholders,
    _soon,
    _today,
)


BOARD_STATUS_LABELS = {
    "overdue": "超期",
    "blocked_risk": "阻塞/风险",
    "pending": "待确认",
    "rework": "返工",
    "in_progress": "进行中",
    "pending_start": "待启动",
    "inq": "前期支持",
    "closed": "已关闭",
}


def list_workbench_board(conn: sqlite3.Connection, query: dict[str, list[str]], user: dict | None = None) -> dict:
    projects = _board_project_rows(conn, query)
    projects = _enrich_project_summaries(conn, projects)
    project_ids = [project["id"] for project in projects]

    task_status = _task_status_stats_by_project(conn, project_ids)
    pending_counts = _pending_counts_by_project(conn, project_ids)
    logs_by_project = _recent_logs_by_project(conn, project_ids)
    owner = (query.get("owner", [""])[0] or "").strip()
    view = (query.get("view", [""])[0] or "all").strip()

    board_projects = []
    for project in projects:
        project_id = project["id"]
        project.update(task_status.get(project_id, {}))
        project.update(pending_counts.get(project_id, {}))
        _set_count_defaults(project)
        project["pending_total"] = (
            project["pending_deliverables"]
            + project["pending_completions"]
            + project["pending_due_date_requests"]
            + project["pending_risk_reviews"]
        )
        project["board_flags"] = _board_flags(project)
        project["board_status"] = _board_status(project)
        project["board_status_label"] = BOARD_STATUS_LABELS[project["board_status"]]
        project["board_group"] = _board_group(project)
        project["board_group_label"] = _board_group_label(project["board_group"])
        project["board_priority"] = _board_priority(project)
        project["recent_logs"] = logs_by_project.get(project_id, [])
        board_projects.append(project)

    board_projects = _filter_board_projects(board_projects, view, owner)
    board_projects.sort(
        key=lambda project: (
            project["board_priority"],
            project.get("current_due_date") or project.get("expected_delivery_date") or "9999-12-31",
            project.get("current_number") or "",
        )
    )

    return {
        "kpis": _board_kpis(board_projects),
        "projects": board_projects,
        "groups": _board_groups(board_projects),
        "current_user": {
            "display_name": (user or {}).get("display_name") or "",
            "roles": (user or {}).get("roles") or [],
        },
    }


def _board_project_rows(conn: sqlite3.Connection, query: dict[str, list[str]]) -> list[dict]:
    filters = ["COALESCE(p.is_deleted, 0) = 0"]
    params: list[str] = []
    search = (query.get("search", [""])[0] or "").strip()
    if search:
        like = f"%{search}%"
        filters.append(
            """
            (
              p.intake_no LIKE ?
              OR p.equipment_no LIKE ?
              OR p.equipment_name LIKE ?
              OR p.project_name LIKE ?
              OR p.project_nature LIKE ?
              OR c.name LIKE ?
              OR cg.name LIKE ?
              OR cs.name LIKE ?
              OR pg.name LIKE ?
              OR co.name LIKE ?
            )
            """
        )
        params.extend([like, like, like, like, like, like, like, like, like, like])

    where = "WHERE " + " AND ".join(filters)
    return [
        row_to_dict(row)
        for row in conn.execute(
            f"""
            SELECT p.id, p.intake_no, p.equipment_no, p.equipment_name, p.project_name,
              p.project_nature, p.project_group_id, p.status_code, s.name AS status_name,
              p.expected_delivery_date, p.project_folder_path, p.created_at, p.updated_at,
              c.name AS customer_name, cg.name AS customer_group_name,
              cs.name AS site_name, pg.name AS project_group_name, co.name AS contact_name
            FROM projects p
            JOIN customers c ON c.id = p.customer_id
            LEFT JOIN customer_groups cg ON cg.id = p.customer_group_id
            LEFT JOIN customer_sites cs ON cs.id = p.site_id
            LEFT JOIN project_groups pg ON pg.id = p.project_group_id
            LEFT JOIN contacts co ON co.id = p.contact_id
            JOIN project_statuses s ON s.code = p.status_code
            {where}
            ORDER BY p.created_at DESC
            LIMIT 500
            """,
            params,
        )
    ]


def _task_status_stats_by_project(conn: sqlite3.Connection, project_ids: list[str]) -> dict[str, dict]:
    if not project_ids:
        return {}
    done_sql = _done_sql()
    placeholders = _placeholders(project_ids)
    rows = conn.execute(
        f"""
        SELECT
          project_id,
          COALESCE(SUM(CASE WHEN status = 'rework' THEN 1 ELSE 0 END), 0) AS rework_tasks,
          COALESCE(SUM(CASE WHEN status = 'in_progress' THEN 1 ELSE 0 END), 0) AS in_progress_tasks,
          COALESCE(SUM(CASE WHEN status = 'waiting_info' THEN 1 ELSE 0 END), 0) AS waiting_info_tasks,
          COALESCE(SUM(CASE WHEN due_date >= ? AND due_date <= ? AND status NOT IN ({done_sql}) THEN 1 ELSE 0 END), 0) AS due_soon_tasks
        FROM execution_tasks
        WHERE project_id IN ({placeholders})
        GROUP BY project_id
        """,
        [_today(), _soon(), *project_ids],
    ).fetchall()
    return {row["project_id"]: row_to_dict(row) for row in rows}


def _pending_counts_by_project(conn: sqlite3.Connection, project_ids: list[str]) -> dict[str, dict]:
    counts = {
        project_id: {
            "pending_deliverables": 0,
            "pending_completions": 0,
            "pending_due_date_requests": 0,
            "pending_risk_reviews": 0,
        }
        for project_id in project_ids
    }
    if not project_ids:
        return counts
    placeholders = _placeholders(project_ids)
    specs = [
        (
            "pending_deliverables",
            f"""
            SELECT project_id, COUNT(*) AS count
            FROM task_deliverables
            WHERE project_id IN ({placeholders}) AND status = 'submitted'
            GROUP BY project_id
            """,
        ),
        (
            "pending_completions",
            f"""
            SELECT project_id, COUNT(*) AS count
            FROM execution_tasks
            WHERE project_id IN ({placeholders}) AND status = 'submitted' AND requires_deliverable = 0
            GROUP BY project_id
            """,
        ),
        (
            "pending_due_date_requests",
            f"""
            SELECT project_id, COUNT(*) AS count
            FROM due_date_change_requests
            WHERE project_id IN ({placeholders}) AND status = 'pending'
            GROUP BY project_id
            """,
        ),
        (
            "pending_risk_reviews",
            f"""
            SELECT project_id, COUNT(*) AS count
            FROM execution_issues
            WHERE project_id IN ({placeholders}) AND status = 'resolved'
            GROUP BY project_id
            """,
        ),
    ]
    for field, sql in specs:
        for row in conn.execute(sql, project_ids):
            counts.setdefault(row["project_id"], {})[field] = int(row["count"] or 0)
    return counts


def _recent_logs_by_project(conn: sqlite3.Connection, project_ids: list[str]) -> dict[str, list[dict]]:
    if not project_ids:
        return {}
    placeholders = _placeholders(project_ids)
    rows = conn.execute(
        f"""
        SELECT project_id, activity_type, title, detail, created_at
        FROM execution_activity_logs
        WHERE project_id IN ({placeholders})
        ORDER BY project_id, created_at DESC
        """,
        project_ids,
    ).fetchall()
    logs: dict[str, list[dict]] = {}
    for row in rows:
        bucket = logs.setdefault(row["project_id"], [])
        if len(bucket) < 5:
            bucket.append(row_to_dict(row))
    return logs


def _set_count_defaults(project: dict) -> None:
    for key in (
        "task_total",
        "task_done",
        "blocked_tasks",
        "submitted_tasks",
        "overdue_tasks",
        "issue_total",
        "open_issues",
        "high_issues",
        "rework_tasks",
        "in_progress_tasks",
        "waiting_info_tasks",
        "due_soon_tasks",
        "pending_deliverables",
        "pending_completions",
        "pending_due_date_requests",
        "pending_risk_reviews",
    ):
        project[key] = int(project.get(key) or 0)


def _board_flags(project: dict) -> list[str]:
    flags = []
    if project["overdue_tasks"]:
        flags.append("超期")
    if project["blocked_tasks"] or project["waiting_info_tasks"]:
        flags.append("阻塞")
    if project["high_issues"]:
        flags.append("高风险")
    if project["pending_total"]:
        flags.append("待确认")
    if project["rework_tasks"]:
        flags.append("返工")
    if project.get("workbench_area") == "inq":
        flags.append("待WO")
    return flags


def _board_status(project: dict) -> str:
    if project.get("workbench_area") == "closed":
        return "closed"
    if project["overdue_tasks"]:
        return "overdue"
    if project["blocked_tasks"] or project["waiting_info_tasks"] or project["high_issues"]:
        return "blocked_risk"
    if project["pending_total"]:
        return "pending"
    if project["rework_tasks"]:
        return "rework"
    if project["in_progress_tasks"]:
        return "in_progress"
    if project.get("workbench_area") == "wo" and not project["task_total"]:
        return "pending_start"
    if project.get("workbench_area") == "inq":
        return "inq"
    return "in_progress"


def _board_priority(project: dict) -> int:
    return {
        "overdue": 0,
        "blocked_risk": 1,
        "pending": 2,
        "rework": 3,
        "in_progress": 4,
        "pending_start": 5,
        "inq": 6,
        "closed": 7,
    }.get(project["board_status"], 8)


def _board_group(project: dict) -> str:
    if project["board_status"] in {"overdue", "blocked_risk", "pending", "rework"}:
        return "attention"
    if project.get("workbench_area") == "closed":
        return "closed"
    if project.get("workbench_area") == "inq":
        return "inq"
    return "normal"


def _board_group_label(group: str) -> str:
    return {
        "attention": "需要处理",
        "normal": "正常推进",
        "inq": "前期支持",
        "closed": "已关闭 / 低关注",
    }.get(group, group)


def _filter_board_projects(projects: list[dict], view: str, owner: str) -> list[dict]:
    if view in {"", "all"}:
        return projects
    if view == "attention":
        return [project for project in projects if project["board_group"] == "attention"]
    if view == "due_soon":
        return [project for project in projects if project["due_soon_tasks"]]
    if view == "overdue":
        return [project for project in projects if project["overdue_tasks"]]
    if view == "blocked":
        return [
            project
            for project in projects
            if project["blocked_tasks"] or project["waiting_info_tasks"] or project["high_issues"] or project["open_issues"]
        ]
    if view == "pending":
        return [project for project in projects if project["pending_total"]]
    if view == "mine":
        if not owner:
            return []
        return [
            project
            for project in projects
            if owner.lower() in (project.get("current_owner") or "").lower()
        ]
    if view == "inq":
        return [project for project in projects if project.get("workbench_area") == "inq"]
    if view == "wo":
        return [project for project in projects if project.get("workbench_area") == "wo"]
    return projects


def _board_kpis(projects: list[dict]) -> dict:
    return {
        "active_projects": sum(1 for project in projects if project.get("workbench_area") != "closed"),
        "due_soon_tasks": sum(project["due_soon_tasks"] for project in projects),
        "overdue_tasks": sum(project["overdue_tasks"] for project in projects),
        "blocked_projects": sum(1 for project in projects if project["blocked_tasks"] or project["waiting_info_tasks"]),
        "pending_confirmations": sum(project["pending_total"] for project in projects),
        "high_risk_projects": sum(1 for project in projects if project["high_issues"]),
    }


def _board_groups(projects: list[dict]) -> list[dict]:
    groups = []
    for key in ("attention", "normal", "inq", "closed"):
        items = [project for project in projects if project["board_group"] == key]
        groups.append({"key": key, "label": _board_group_label(key), "count": len(items)})
    return groups

File: static/js/workbench-board.js
const BOARD_KPI_DEFINITIONS = [
  { key: "active_projects", label: "进行中项目", view: "all" },
  { key: "due_soon_tasks", label: "本周到期", view: "due_soon" },
  { key: "overdue_tasks", label: "超期任务", view: "overdue" },
  { key: "blocked_projects", label: "阻塞项目", view: "blocked" },
  { key: "pending_confirmations", label: "待确认", view: "pending" },
  { key: "high_risk_projects", label: "高风险", view: "blocked" },
];

const BOARD_STATUS_CLASS = {
  overdue: "danger",
  blocked_risk: "danger",
  pending: "warn",
  rework: "warn",
  in_progress: "ok",
  pending_start: "neutral",
  inq: "neutral",
  closed: "muted",
};

function boardOwnerName() {
  return state.auth.user?.display_name || state.auth.user?.email?.split("@")[0] || $("#workbenchOwnerInput")?.value.trim() || "";
}

function boardQueryParams() {
  const params = new URLSearchParams();
  const search = $("#boardSearchInput").value.trim();
  const view = state.boardFilter || "all";
  if (search) params.set("search", search);
  if (view) params.set("view", view);
  if (view === "mine") params.set("owner", boardOwnerName());
  return params;
}

async function loadProjectBoard(selectProjectId = state.boardSelectedProjectId) {
  updateBoardFilterButtons();
  const payload = await api(`/api/workbench/board?${boardQueryParams().toString()}`);
  state.boardPayload = payload;
  state.boardProjects = payload.projects || [];
  renderBoardKpis(payload.kpis || {});
  renderBoardProjectList(selectProjectId);
  const selectedExists = state.boardProjects.some((project) => project.id === selectProjectId);
  const selectedId = selectedExists ? selectProjectId : state.boardProjects[0]?.id || null;
  renderBoardSnapshot(selectedId);
}

function updateBoardFilterButtons() {
  const filter = state.boardFilter || "all";
  document.querySelectorAll("[data-board-filter]").forEach((button) => {
    button.classList.toggle("active", button.dataset.boardFilter === filter);
  });
}

function renderBoardKpis(kpis = {}) {
  $("#boardKpis").innerHTML = BOARD_KPI_DEFINITIONS.map((item) => `
    <button type="button" class="board-kpi" data-board-kpi-view="${escapeHtml(item.view)}">
      <span>${escapeHtml(kpis[item.key] || 0)}</span>
      <small>${escapeHtml(item.label)}</small>
    </button>
  `).join("");
  $("#boardKpis").querySelectorAll("[data-board-kpi-view]").forEach((button) => {
    button.addEventListener("click", () => {
      state.boardFilter = button.dataset.boardKpiView || "all";
      loadProjectBoard().catch(console.error);
    });
  });
}

function renderBoardProjectList(selectedId = state.boardSelectedProjectId) {
  const list = $("#boardProjectList");
  if (!state.boardProjects.length) {
    list.innerHTML = `<div class="empty">暂无匹配项目</div>`;
    return;
  }
  const groups = state.boardPayload?.groups || [];
  list.innerHTML = groups
    .map((group) => {
      const projects = state.boardProjects.filter((project) => project.board_group === group.key);
      if (!projects.length) return "";
      return `
        <section class="board-group">
          <div class="board-group-head">
            <h3>${escapeHtml(group.label)}</h3>
            <span>${escapeHtml(projects.length)} 项</span>
          </div>
          <div class="board-group-list">
            ${projects.map((project) => renderBoardProjectRow(project, selectedId)).join("")}
          </div>
        </section>
      `;
    })
    .join("");
  list.querySelectorAll("[data-board-project-id]").forEach((button) => {
    button.addEventListener("click", () => {
      renderBoardSnapshot(button.dataset.boardProjectId);
    });
  });
}

function renderBoardProjectRow(project, selectedId) {
  const active = project.id === selectedId ? " active" : "";
  const due = project.current_due_date || project.expected_delivery_date || "";
  const statusClass = BOARD_STATUS_CLASS[project.board_status] || "neutral";
  const progress = project.task_total ? `${project.task_done || 0}/${project.task_total}` : "未建任务";
  const flags = (project.board_flags || []).map((flag) => `<span class="tag ${boardFlagClass(flag)}">${escapeHtml(flag)}</span>`).join("");
  return `
    <button type="button" class="board-project-row${active}" data-board-project-id="${escapeHtml(project.id)}">
      <span class="board-row-number">
        <strong>${escapeHtml(project.current_number || project.intake_no || "")}</strong>
        <small>${escapeHtml(workbenchAreaName(project.workbench_area))}</small>
      </span>
      <span class="board-row-main">
        <strong>${escapeHtml(project.equipment_name || project.project_name || "")}</strong>
        <small>${escapeHtml(project.customer_name || "")}${project.site_name ? ` · ${escapeHtml(project.site_name)}` : ""}${project.project_group_name ? ` · ${escapeHtml(project.project_group_name)}` : ""}</small>
      </span>
      <span class="board-row-stage">
        <span class="status-pill ${statusClass}">${escapeHtml(project.board_status_label)}</span>
        <small>${escapeHtml(project.status_name || "")}</small>
      </span>
      <span class="board-row-owner">
        <strong>${escapeHtml(project.current_owner || "未指定")}</strong>
        <small>责任人</small>
      </span>
      <span class="board-row-next">
        <strong>${escapeHtml(project.next_action || "未设置下一步")}</strong>
        <small>${escapeHtml(progress)} 任务</small>
      </span>
      <span class="board-row-due">
        <strong class="${dueClass(due)}">${escapeHtml(due || "未设")}</strong>
        <small>Due Date</small>
      </span>
      <span class="board-row-flags">${flags || `<span class="tag neutral">正常</span>`}</span>
    </button>
  `;
}

function boardFlagClass(flag) {
  if (["超期", "高风险"].includes(flag)) return "danger";
  if (["阻塞", "返工", "待WO"].includes(flag)) return "warn";
  if (flag === "待确认") return "";
  return "neutral";
}

function renderBoardSnapshot(projectId) {
  state.boardSelectedProjectId = projectId || null;
  renderBoardProjectList(projectId);
  const project = state.boardProjects.find((item) => item.id === projectId);
  if (!project) {
    $("#boardSnapshot").innerHTML = `<div class="empty">请选择一个项目查看快照</div>`;
    return;
  }
  const canOpenLibrary = userHasRole("pm");
  const pendingRows = [
    ["交付文件", project.pending_deliverables],
    ["完成说明", project.pending_completions],
    ["改期申请", project.pending_due_date_requests],
    ["风险关闭", project.pending_risk_reviews],
  ];
  $("#boardSnapshot").innerHTML = `
    <div class="board-snapshot-head">
      <div>
        <h3>${escapeHtml(project.current_number || project.intake_no || "")}</h3>
        <p>${escapeHtml(project.equipment_name || project.project_name || "")}</p>
      </div>
      <span class="status-pill ${BOARD_STATUS_CLASS[project.board_status] || "neutral"}">${escapeHtml(project.board_status_label)}</span>
    </div>
    <dl class="board-snapshot-grid">
      <dt>客户/工厂</dt>
      <dd>${escapeHtml(project.customer_name || "")}${project.site_name ? ` · ${escapeHtml(project.site_name)}` : ""}</dd>
      <dt>产品/产线</dt>
      <dd>${escapeHtml(project.project_group_name || "未关联")}</dd>
      <dt>当前阶段</dt>
      <dd>${escapeHtml(project.status_name || workbenchAreaName(project.workbench_area))}</dd>
      <dt>当前责任人</dt>
      <dd>${escapeHtml(project.current_owner || "未指定")}</dd>
      <dt>下一步动作</dt>
      <dd>${escapeHtml(project.next_action || "未设置")}</dd>
      <dt>Due Date</dt>
      <dd class="${dueClass(project.current_due_date || project.expected_delivery_date || "")}">${escapeHtml(project.current_due_date || project.expected_delivery_date || "未设置")}</dd>
    </dl>
    <div class="board-snapshot-metrics">
      <div><span>${escapeHtml(project.task_done || 0)}/${escapeHtml(project.task_total || 0)}</span><small>任务</small></div>
      <div><span>${escapeHtml(project.open_issues || 0)}</span><small>打开风险</small></div>
      <div><span>${escapeHtml(project.pending_total || 0)}</span><small>待确认</small></div>
    </div>
    <section class="board-snapshot-section">
      <h4>待处理摘要</h4>
      <div class="board-pending-list">
        ${pendingRows.map(([label, count]) => `
          <span class="${count ? "active" : ""}">${escapeHtml(label)} <strong>${escapeHtml(count || 0)}</strong></span>
        `).join("")}
      </div>
    </section>
    <section class="board-snapshot-section">
      <h4>最近日志</h4>
      <div class="board-log-list">
        ${(project.recent_logs || []).length ? project.recent_logs.map((log) => `
          <div>
            <strong>${escapeHtml(log.title || "")}</strong>
            <small>${escapeHtml(workbenchDateValue(log.created_at || ""))}${log.detail ? ` · ${escapeHtml(log.detail)}` : ""}</small>
          </div>
        `).join("") : `<div class="empty small-empty">暂无执行日志</div>`}
      </div>
    </section>
    <div class="board-snapshot-actions">
      <button type="button" data-board-action="open-workbench" data-project-id="${escapeHtml(project.id)}">进入项目执行</button>
      ${canOpenLibrary ? `<button type="button" class="secondary" data-board-action="open-library" data-project-id="${escapeHtml(project.id)}">资料库详情</button>` : ""}
    </div>
  `;
  bindBoardSnapshotActions();
}

function bindBoardSnapshotActions() {
  $("#boardSnapshot").querySelectorAll("[data-board-action]").forEach((button) => {
    button.addEventListener("click", async () => {
      const projectId = button.dataset.projectId;
      try {
        if (button.dataset.boardAction === "open-workbench") {
          switchView("workbench", false);
          await loadWorkbenchProjects(projectId);
        }
        if (button.dataset.boardAction === "open-library") {
          await openDetail(projectId);
        }
      } catch (error) {
        showToast(error.message);
      }
    });
  });
}
